docs/rfcs/038-independent-compute-release.md

```diff
@@ -0,0 +1,320 @@
+
+In the code:
+
+- [Release](https://github.com/helm/helm/blob/2b30cf4b61d587d3f7594102bb202b787b9918db/pkg/release/release.go#L20-L43)
+- [Release info](https://github.com/helm/helm/blob/2b30cf4b61d587d3f7594102bb202b787b9918db/pkg/release/info.go#L24-L40)
+- [Release status](https://github.com/helm/helm/blob/2b30cf4b61d587d3f7594102bb202b787b9918db/pkg/release/status.go#L18-L42)
+
+TL;DR it has several important attributes:
+
+- Revision -- unique release ID/primary key. It is not the same as the application version,
+  because the same version can be deployed several times, e.g., after a newer version rollback.
+- App version -- version of the application chart/code.
+- Config -- set of overrides to the default config of the application.
+- Status -- current status of the release in the history.
+- Timestamps -- tracks when a release was created and deployed.
+
+## Proposed implementation
+
+### Separate release branch
+
+We will use a separate release branch, `release-compute`, to have a clean history for releases and commits.
+In order to avoid confusion with storage releases, we will use a different prefix for compute [git release
+tags](https://github.com/neondatabase/neon/releases) -- `release-compute-XXXX`. We will use the same tag for
+Docker images as well. The `neondatabase/compute-node-v16:release-compute-XXXX` looks longer and a bit redundant,
+but it's better to have image and git tags in sync.
+
+### Compute config/settings manifest
+
+We will create a new sub-directory `compute` and file `compute/manifest.yaml` with a structure:
+
+```yaml
+pg_settings:
+  # Common settings for primaries and secondaries of all versions.
+  common:
+    wal_log_hints: "off"
+    max_wal_size: "1024"
+
+  per_version:
+    14:
+      # Common settings for both replica and primary of version PG 14
+      common:
+        shared_preload_libraries: "neon,pg_stat_statements,extension_x"
+    15:
+      common:
+        shared_preload_libraries: "neon,pg_stat_statements,extension_x"
+      # Settings that should be applied only to
+      replica:
+        # Available only starting Postgres 15th
+        recovery_prefetch: "off"
+    # ...
+    17:
+      common:
+        # For example, if third-party `extension_x` is not yet available for PG 17
+        shared_preload_libraries: "neon,pg_stat_statements"
+      replica:
+        recovery_prefetch: "off"
+```
+
+**N.B.** Setting value should be a string with `on|off` for booleans and a number (as a string)
+without units for all numeric settings. That's how the control plane currently operates.
+
+The priority of settings will be (a higher number is a higher priority):
+
+1. Any static and hard-coded settings in the control plane
+2. `pg_settings->common`
+3. Per-version `common`
+4. Per-version `replica`
+5. Any per-user/project/endpoint overrides in the control plane
+6. Any dynamic setting calculated based on the compute size
```

Comment:
Will there evere be nested settings? Then we should specify the merge strategy here, or, explicitly call out how we'll implement more complicated merge strategies in the future.

I'd love if we can rule out nesting completely to avoid this complexity. Quickly gets into jsonpatch territory.

Example: what if `shared_preload_libraries` were a list:

```
shared_preload_libraries:
  - neon
  - pg_stat_statements  
```

What happens if two levels in the priority hierarchy specify `shared_preload_libraries`?
Does the higher priority level replace, or does it merge?

Reply:
Hm, I think for custom extensions we definitely want a merge strategy.

Reply:
Yeah, `shared_preload_libraries` is probably the only exclusion. Right now, we do not do a 'set merge' in the control plane either, where we have different levels of overrides already. If it's an override, it's a complete override, i.e., we just enumerate all 'base' extensions + custom one on the user or project level. I tend to keep it like that for now because, at each level, you fully know what's the override. Otherwise, it will be tricky and a bit scattered; but it's a fair point

Reply:
Added a note about `shared_preload_libraries` in the https://github.com/neondatabase/neon/pull/8881/commits/1939b99ec1038531064a02d8779d53301ce33f29